llama-index-experimental/llama_index/experimental/embedding_finetuning/base.py

```diff
@@ -0,0 +1,172 @@
+import logging
+from typing import Any, Optional
+
+import torch
+from torch import nn, Tensor
+from torch.nn import functional as F
+from torch.utils.data import DataLoader
+from tqdm import tqdm, trange
+
+from llama_index.core.utils import print_text
+from llama_index.core.base.embeddings.base import BaseEmbedding
+from llama_index.core.utils import infer_torch_device
+from llama_index.finetuning.embeddings.common import EmbeddingQAFinetuneDataset
+
+logger = logging.getLogger(__name__)
+
+
+def multiclass_nll_loss(output, targets):
+    return (-1 * output * targets).sum(axis=-1).mean()
+
+
+class EmbeddingFinetuneEngine:
```

Comment:
given method is specifically `NUDGE`, naming here is too generic?

Reply:
Thanks for renaming!